### PR TITLE
Add direct reading capability for ADIOS IO type

### PR DESCRIPTION
**Pull request by dqwu**
This enhancement is primarily authored by @dmitry-ganyushin (Dmitry
Ganyushin, ADIOS developer), which enables E3SM runs to directly
read restart files in ADIOS BP5 format, eliminating the need for
BP to NC file conversion before restart runs.

Furthermore, we have incorporated additional code refinements,
improved error handling, and addressed minor bugs in Dmitry's
initial patch.

### CODE REVIEW

**Comment by dqwu** [comment]
[Two Known Limitations of ADIOS Read Support]
1) In order to read a distributed array, the decomposition map must
precisely match the map utilized for writing that array. This also
requires that the restart run employs the same PE layout as the
initial run.

For E3SM ne4 F case restart runs, we encounter some errors when
attempting to read the restart file mpassi.rst.0001-01-02_00000.nc
in ADIOS BP format. This arises from certain distributed arrays
being written and subsequently read with differing decomposition
maps.

A viable solution is to utilize NetCDF or PnetCDF types for writing
and reading this category of restart files.

2) Currently, open-to-append mode is not supported for ADIOS type.
If an E3SM restart run is configured to open an output history file
in ADIOS BP format and append additional time steps to it, an error
will be generated by SCORPIO.

Workaround 1: Adjust the configuration of restart runs to avoid
open-to-append mode for specific history files.

Workaround 2: Opt for NetCDF or PnetCDF types for files that will
undergo append operations.

**Comment by dqwu** [comment]
[Known Issues on E3SM/CIME Side]
1) In E3SM restart runs, it's necessary to create dummy XXX.nc files
even when XXX.nc.bp files are already available. This workaround is
crucial to successfully pass certain file existence checks on the
E3SM/CIME side. Ideally, E3SM/CIME should be able to recognize
these files in ADIOS BP format.

2) Some MPAS files are written and read with IO types specified in
stream files, rather than in env_run.xml. Currently, we're still
encountering problems when attempting to set ADIOS IO types in
these stream files, resulting in the consistent use of the default
PnetCDF type.

**Review by tkurc — APPROVED** [review]

**Review by jayeshkrishna — COMMENTED** [review]
Do we need qhashtbl* files? What is the copyright for this code (does it work with SCORPIO copyright)? Can we replace it with C++ maps?

File: src/clib/pio_darray.c [hunk 1]
```diff
@@ -2548,6 +2548,843 @@ if (file->iotype != PIO_IOTYPE_HDF5)
...
+    if (att_name == NULL)
+    {
+        return pio_err(NULL, file, PIO_ENOMEM, __FILE__, __LINE__,
+                       "Reading variable (%s, varid=%d) from file (%s, ncid=%d) using ADIOS iotype failed. "
+                       "Out of memory allocating %lld bytes for attribute name",
+                       pio_get_vname_from_file(file, varid), varid, pio_get_fname_from_file(file), file->pio_ncid, (long long int) att_name_len);
+    }
+
+    strncpy(att_name, prefix_var_name, att_name_len);
+    strcat(att_name, adios_vdesc->name);
```
[the remark this answers is not in the document]
Comment (dqwu) [reply src/clib/pio_darray.c:2722]: We can do this in a future PR to replace all existing strcats by strncats.

**Review by jayeshkrishna — APPROVED** [review]

**Pull request merged**